[DashTree] Timeline improvements for live streams

CastagnaIT opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
While investigating to fix issue i noticed various problems, some due to partial implementations that were never completed

in short
- implementation of TSB timeline update between periods for MPD that use SegmentTemplate without a defined TimeLine,
the code relative to TSB timeline update, will check also for expired segments that fall outside the TSB, then delete them all in realtime, this fix the #1769 issue because previously expired segments was never deleted and the video seek was pointing always to the same (old) segment, with the problem that was impossible seek back to "live"
- implementation of the removal of expired periods, where previously they were persistently kept in memory, with the side effect of a small "memory leak" as well
- add dynamic update of period/chapter duration, this will allow the passage of time to be displayed appropriately in the GUI seek bar
- Reverted dash segment-controlled live stream support (implemented by #1525), this is now not more needed since segments for templated representations are now inserted while in playback without any restrictions. This use case has been tested by user see comments below

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
fix #1769 

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Live Dash with segment templated representations (no defined segments)
example on ISASamples addon test stream "dashif periods_20 multi-period"
- tested also live HLS having long timeshift

some users also tested

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the Wiki documentation
- [ ] I have updated the documentation accordingly

CastagnaIT committed:
[Session] Prefer video stream on m_timingStream

CastagnaIT committed:
Changes to current segment and adaptations

CastagnaIT committed:
[DashTree] Removed InsertLiveFragment implementation

InsertLiveSegment now can be used to do the same thing

CastagnaIT commented:
@mtr81 can you help me to do a test?
time ago you have opened this issue #1495, for a specific live stream that was not working,
have you the possibility to test that same live stream to see if it works with this PR? (kodi 22)

Test builds: https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.adaptive/detail/PR-1867/3/artifacts

mtr81 commented:
Hi
Playback stops after several seconds.
[log.txt](https://github.com/user-attachments/files/21200872/log.txt)

CastagnaIT commented:
@mtr81 i made a fix, new build:
https://jenkins.kodi.tv/blue/organizations/jenkins/xbmc%2Finputstream.adaptive/detail/PR-1867/4/artifacts

mtr81 commented:
Works fine. Returning to live mode after rewinding to previous segments also works as expected.

CastagnaIT commented:
perfect thanks

CastagnaIT committed:
[DashTree] Timeline improvements for live streams

This change implement
- TSB timeline update between periods for SegmentTemplate without a defined timeline
- Removal of expired periods
- Update in realtime of period/chapter duration so that GUI seekbar show
  correctly the passage of time

CastagnaIT commented on src/Session.cpp:
```diff
@@ -719,6 +719,15 @@ void CSession::EnableStream(CStream* stream, bool enable)
   }
 }
 
+uint64_t SESSION::CSession::GetTotalTimeMs() const
+{
+  // In live streaming do not take into account the live delay duration, because its not seekable
+  if (m_adaptiveTree->IsLive() && m_adaptiveTree->m_totalTime > m_adaptiveTree->m_liveDelay * 1000)
+    return m_adaptiveTree->m_totalTime - m_adaptiveTree->m_liveDelay * 1000;
+  else
+    return m_adaptiveTree->m_totalTime;
+}
```
PS: I wrote that the live delay part should not take into account, this is to prevent the user from attempting to seek in this part of the video (being the GetTotalTimeMs method is involved to the GUI bar to shown the duration of the video). This is because if you attempt to seek in to live delay part, ISA try limits this attempt on CSession::SeekTime, but Kodi core does not care about this limit and Kodi core keeps dropping packets until it gets the PTS it expects on packets, this can stall playback in buffering for several seconds and should be avoided
